feat: [dpDayPicker] directive

isaacplmann opened the pull request:
Implements #69 .

Handles all config options that `<dp-day-picker>` does except:
1. ~minDate~
2. ~maxDate~
3. showGoToCurrent
4. ~api.open()~
5. ~api.close()~

Basic syntax is this:
```html
<md-input-container>
  <input mdInput [dpDayPicker]="config" name="datePicker"
         theme="dp-material" [(ngModel)]="date" />
</md-input-container>
```

There's also an optional `inputElementContainer` input on the `dpDayPicker` directive.  If unspecified it will start at the `dpDayPicker`'s host component and keep going up to parent elements until it finds an element that matches `.mat-input-wrapper` (the class that `@angular/material` uses).  `inputElementContainer` can be a string (which will be an argument for `document.querySelector` to find the appropriate element) or an `ElementRef` which will be used to directly.

I'm not sure if I quite like the way I'm handling `inputElementContainer`, but it's working well enough to show you for feedback.

FYI, you could also use a plain `<input>` like this:
```html
<input [dpDayPicker]="config" name="datePicker"
       id="inputContainer" inputElementContainer="#inputContainer"
       theme="dp-material" [(ngModel)]="date" />
```

vlio20 commented on src/app/date-picker/date-picker.component.ts:
```diff
@@ -233,7 +233,13 @@ export class DatePickerComponent implements OnChanges,
   setElementPositionInDom() {
     this.calendarWrapper = <HTMLElement> this.calendarContainer.nativeElement;
-    this.inputElementContainer = this.elemRef.nativeElement.querySelector('.dp-input-container');
+    let inputElementContainer = this.componentConfig.inputElementContainer;
```
pls move to separate method  all added logic in this method

vlio20 commented on src/app/date-picker/date-picker.directive.ts:
```diff
@@ -0,0 +1,82 @@
+import {DatePickerComponent} from './date-picker.component';
+import {OnInit} from '@angular/core';
+import {ViewContainerRef} from '@angular/core';
+import { Directive, ComponentFactoryResolver, Input, HostListener } from '@angular/core';
```
remove spaces

vlio20 commented on src/app/date-picker/date-picker.directive.ts:
```diff
@@ -0,0 +1,82 @@
+import {OnInit} from '@angular/core';
+import {ViewContainerRef} from '@angular/core';
+import { Directive, ComponentFactoryResolver, Input, HostListener } from '@angular/core';
+import { Moment } from 'moment';
```
same

vlio20 commented on src/app/date-picker/date-picker.directive.ts:
```diff
@@ -0,0 +1,82 @@
+  ) { }
+
+  ngOnInit(): void {
+    const factory = this.componentFactoryResolver.resolveComponentFactory(DatePickerComponent);
```
pls move this logic to separate method , and it would be a good idea to break it to other methods. setting the elements, setting the validation and register listeners.

vlio20 commented on src/app/date-picker/date-picker.directive.ts:
```diff
@@ -0,0 +1,82 @@
+  @Input() inputElementContainer: ElementRef;
+  @Input() theme: string;
+
+  public datePicker: DatePickerComponent;
```
pls create a service for this directive.

vlio20 commented on src/app/date-picker/date-picker.directive.ts:
```diff
@@ -0,0 +1,82 @@
+  ngOnChanges(changes: SimpleChanges) {
+    const { config, theme, inputElementContainer } = changes;
+    if (this.datePicker) {
+      if (theme && theme.currentValue) {
```
please see how this was handled in the date-picker component - `@HostBinding('class') @Input() theme: string;`

isaacplmann replied:
I can't bind to the class of the host component - I need to set the theme on the datePicker component.  I can do it this way or using a setter on the `theme` input - let me know if you prefer the setter method.

vlio20 replied:
pls use a setter

vlio20 commented on src/app/date-picker/date-picker.directive.ts:
```diff
@@ -0,0 +1,82 @@
+    } else if (this.inputElementContainer) {
+      config.inputElementContainer = this.inputElementContainer.nativeElement;
+    } else {
+      config.inputElementContainer = this.utils.closestParent(this.viewContainerRef.element.nativeElement, '.mat-input-wrapper');
```
what is `mat-input-wrapper`?

isaacplmann replied:
`mat-input-wrapper` is a class specific to `@angular/material`'s implementation of their `mdInput` directive.  I probably shouldn't make `ng2-date-picker` directly coupled to their implementation.  I'll remove that default.

vlio20 replied:
what selector you will pass here than?

isaacplmann replied:
I'll default to the host component element.

vlio20 replied:
ok, should it be configurable?

vlio20 commented on src/app/date-picker/date-picker.directive.ts:
```diff
@@ -0,0 +1,82 @@
+    this.datePicker.onClick();
+  }
+
+  private updateDatepickerConfig(config: IDatePickerConfig): IDatePickerConfig {
```
move to service

vlio20 replied:
pls create specific interface for this directive's config object

isaacplmann replied:
Made `IDatePickerDirectiveConfig` that is identical to `IDatePickerConfig` except it removes `showGoToCurrent`.  Should I have `IDatePickerConfig` extend `IDatePickerDirectiveConfig`?

vlio20 replied:
nope

vlio20 commented on src/app/common/services/utils/utils.service.ts:
```diff
@@ -153,4 +153,13 @@ export class UtilsService {
     }
   }
+
+  // todo:: add unit test
```
please add the unit test

vlio20 commented on src/app/date-picker/date-picker.directive.ts:
```diff
@@ -0,0 +1,82 @@
+import { Directive, ComponentFactoryResolver, Input, HostListener } from '@angular/core';
+import { Moment } from 'moment';
+
+@Directive({ selector: '[dpDayPicker][ngModel]' })
```
pls break to lines

vlio20 commented on src/app/date-picker/date-picker.directive.ts:
```diff
@@ -0,0 +1,82 @@
+    this.datePicker.registerOnChange(value => {
+      this.model.control.setValue(this.datePicker.inputElementValue, { emitEvent: false });
+    });
+    if (this.config) {
```
this if is redundant, the service should handle cases when there is no config object

vlio20 commented on package.json:
```diff
@@ -36,16 +36,18 @@
+    "@angular/core": "^4.0.3",
+    "@angular/forms": "^4.0.3",
+    "@angular/http": "^4.0.3",
+    "@angular/material": "^2.0.0-beta.3",
```
why do we need this module? I would like to do all our testing on native elements.

isaacplmann replied:
This is just to show people that they can use it with angular material's `mdInput` directive.  That's probably the most likely use case for this directive, especially given the built in material theme.  I can use a plain input if you don't think there's value in the material example.

vlio20 replied:
Please do it with a native input element. I don't want to add the Material module to this project.

isaacplmann replied:
It wouldn't be a dependency to use the library, just to run the demo.  But I'll switch it to a native input anyway.

vlio20 replied:
I know, thanks.

vlio20 commented on src/app/demo/demo/demo.component.html:
```diff
@@ -123,6 +124,24 @@
         </div>
       </div>
+
+      <div class="dp-picker-container" *ngIf="pickerMode === 'dayDirective'">
```
pls add e2e for this feature

vlio20 commented:
@isaacplmann, please also add unit tests, e2e (protractor) tests and documentation for this feature.

isaacplmann commented:
1. I've added unit tests and e2e tests.
2. I also added the ability to set a 'type' on the [dpDayPicker] directive so it can be used as a month picker as well.
3. api is now accessible from the [dpDayPicker] directive

Todo:
[ ] Documentation.
[ ] Switch the ngOnChanges functionality to use setters instead.

isaacplmann commented:
I think I've made all the changes you asked for.  And I've updated the README.md with documentation.  Let me know what you think.

isaacplmann commented on package.json:
```diff
@@ -36,7 +36,7 @@
   },
   "devDependencies": {
-    "@angular/cli": "^1.0.0",
+    "@angular/cli": "1.0.0",
```
FYI, the library doesn't compile with @angular/cli@1.0.1

vlio20 replied:
mmm... can you tell why?

isaacplmann replied:
[Yes](https://github.com/angular/angular-cli/issues/2034#issuecomment-297274534).  I can put in a workaround for the angular-cli bug.  There are two workaround options:

1. Split all interfaces into separate files
-OR-
2. Do this:
```ts
  // tslint:disable-next-line:typedef
  @Input() type = <CalendarType>'day';
```

instead of:
```ts
  @Input() type: CalendarType = 'day';
```


Should I fix this in a separate PR?

vlio20 replied:
I would do the 1 option.
Yes, new PR would be gr8!

vlio20 commented on src/app/date-picker/date-picker-directive.service.ts:
```diff
@@ -4,12 +4,14 @@ import {ElementRef, Injectable} from '@angular/core';
 export class DatePickerDirectiveService {
   getConfig(config: IDatePickerDirectiveConfig = {}, attachTo?: ElementRef | string): IDatePickerDirectiveConfig {
-    config.hideInputContainer = true;
+    const baseConfig = { hideInputContainer: true };
```
please see how the default configuration is set for other components.

vlio20 commented on src/app/date-picker/date-picker.directive.ts:
```diff
@@ -1,28 +1,65 @@
-  };
+export class DatePickerDirective implements OnInit {
+  private _config: IDatePickerDirectiveConfig;
+  private _attachTo: ElementRef | string;
```
I think this should be only `ElementRef`, strings should be handled before the assignment

vlio20 commented:
@isaacplmann, I merged #71. Can you please merge this PR with it?  so it will be possible to merge it via github's ui.

isaacplmann commented:
Merged.  All tests passed.

vlio20 commented:
thanks @isaacplmann for another great feature!
